addon/internal-session.js

```diff
@@ -44,58 +44,86 @@ export default Ember.ObjectProxy.extend(Ember.Evented, {
 
   restore() {
     return new Ember.RSVP.Promise((resolve, reject) => {
-      let restoredContent   = this.store.restore();
-      let { authenticator } = (restoredContent.authenticated || {});
-      if (!!authenticator) {
-        delete restoredContent.authenticated.authenticator;
-        this.container.lookup(authenticator).restore(restoredContent.authenticated).then((content) => {
-          this.set('content', restoredContent);
-          this._setup(authenticator, content);
-          resolve();
-        }, () => {
-          Ember.Logger.debug(`The authenticator "${authenticator}" rejected to restore the session - invalidating…`);
+      this._restoreStoreAsync().then((restoredContent) => {
+        let { authenticator } = (restoredContent.authenticated || {});
+        if (!!authenticator) {
+          delete restoredContent.authenticated.authenticator;
+          this.container.lookup(authenticator).restore(restoredContent.authenticated).then((content) => {
+            this.set('content', restoredContent);
+            this._setup(authenticator, content).then(resolve, reject);
+          }, () => {
+            Ember.Logger.debug(`The authenticator "${authenticator}" rejected to restore the session - invalidating…`);
+            this.set('content', restoredContent);
+            this._clear().then(resolve, reject);
+          });
+        } else {
+          delete (restoredContent || {}).authenticated;
           this.set('content', restoredContent);
-          this._clear();
-          reject();
-        });
-      } else {
-        delete (restoredContent || {}).authenticated;
-        this.set('content', restoredContent);
-        this._clear();
-        reject();
-      }
+          this._clear().then(resolve, reject);
+        }
+      }, () => {
+        this._clear().then(resolve, reject);
+      });
     });
   },
 
+  _restoreStoreAsync() {
+    let restoredContent = this.store.restore();
+    if (typeof restoredContent === 'undefined' || typeof restoredContent.then === 'undefined') {
+      Ember.deprecate('Ember Simple Auth: Synchronous stores have been deprecated. Make sure your custom store\'s `restore` method returns a promise.', false);
+      return Ember.RSVP.Promise.resolve(restoredContent || {});
+    } else {
+      return restoredContent;
+    }
+  },
+
+  _persistStoreAsync(data) {
+    let persisted = this.store.persist(data);
+    if (typeof persisted === 'undefined' || typeof persisted.then === 'undefined') {
+      Ember.deprecate('Ember Simple Auth: Synchronous stores have been deprecated. Make sure your custom store\'s `persist` method returns a promise.', false);
+      return Ember.RSVP.Promise.resolve();
+    } else {
+      return persisted;
+    }
+  },
```

Comment:
Would be better to have one method that handles both as the methods are more or less identical.